src/geo/transform.js

```diff
@@ -569,10 +569,13 @@ class Transform {
         // No change of LOD behavior for pitch lower than 60 and when there is no top padding: return only tile ids from the requested zoom level
         const minZoom = this.pitch <= 60.0 && this._edgeInsets.top <= this._edgeInsets.bottom && !this._elevation ? z : 0;
 
-        const maxRange = this.elevation ? this.elevation.exaggeration() * 10000 : 0;
+        // When calculating tile cover for terrain, create deep AABB for nodes, to ensure they intersect frustum: for sources,
+        // other than DEM, use minimum of visible DEM tiles and center altitude as upper bound (pitch is always less than 90°).
+        const maxRange = options.isTerrainDEM && this._elevation ? this._elevation.exaggeration() * 10000 : this._centerAltitude;
+        const minRange = options.isTerrainDEM ? -maxRange : this._elevation ? this._elevation.getMinElevationBelowMSL() : 0;
```

Comment:
Something to note is when terrain._visibleDemTiles == 0, we may hit a corner case: when there are tiles fully within negative elevation we might miss tiles, because they don't intersect with the frustum (so they are not accounted for in getMinElevationBelowMSL). One approach would be to use with an overly larger frustum for tile cover, to ensure they are marked as visible at least once. It's not directly related to this PR (same happens in main).